Improve parameters parsing and fix callbacks

albestro opened the pull request:
Close #117 

As per discussion in the issue, I tried to improve a bit the parser for parameters. It is far from being perfect.

The new method should is not regex based, it should be able to detect all types and pointers with related constness. The constness information is stored inside `Par` as a list of bools, intrinsically giving two information:
- it is a value (`len==0`), a pointer (`len==1`) or a pointer to pointer (`len==2`)
- constness for each existing level of indirection

e.g.
- `const int *` -> `[True, False]`
- `const int * const` -> `[True, True]`
- `int * const` -> `[False, True]`

albestro commented:
For the moment I worked to be able to reproduce (more or less) the same generated wrappers.

Here ti is a diff between what I generated with master and my branch

```bash
1579c1579
<     MediaReadCb = ctypes.CFUNCTYPE(ctypes.POINTER(ctypes.c_ssize_t), ctypes.c_void_p, ctypes.POINTER(ctypes.c_char), ctypes.c_size_t)
---
>     MediaReadCb = ctypes.CFUNCTYPE(ctypes.POINTER(ctypes.c_ssize_t), ctypes.c_void_p, ctypes.c_char_p, ctypes.c_size_t)
5876c5876
<                     ctypes.c_int, Media, MediaSlaveType, ctypes.c_uint, ctypes.c_char_p)
---
>                     ctypes.c_int, Media, MediaSlaveType, ctypes.c_int, ctypes.c_char_p)
5902c5902
<                     ctypes.c_uint, Media, ctypes.POINTER(ctypes.POINTER(MediaSlave)))
---
>                     ctypes.c_int, Media, ctypes.POINTER(ctypes.POINTER(MediaSlave)))
5913c5913
<                     None, ctypes.POINTER(MediaSlave), ctypes.c_uint)
---
>                     None, ctypes.POINTER(MediaSlave), ctypes.c_int)
7250c7250
<                     None, MediaPlayer, Position, ctypes.c_uint)
---
>                     None, MediaPlayer, Position, ctypes.c_int)
7659c7659
<                     ctypes.c_int, MediaPlayer, ctypes.c_uint, ctypes.c_char_p, ctypes.c_uint, ctypes.c_uint)
---
>                     ctypes.c_int, MediaPlayer, ctypes.c_uint, ctypes.c_char_p, ctypes.c_int, ctypes.c_int)

```

Some highlights:
- some `int` are now recognized as `unsigned int` (added the new type mapping)
- added in `type2class` a "wrong" mapping for `unsigned char*` to not alter too much the wrapper (at least for the moment)
- added in type2class_out the "right" mapping `unsigned char*` to `ctypes.POINTER(ctypes.c_char)` to make the callback `media_read_cb` get the right wrapper function signature

oaubert commented on generator/generate.py:
```diff
@@ -687,31 +686,56 @@ def parse_groups(self, match_t, match_re, ends=';'):
+                    param_type = split_value[0]
+                    param_name = ''
+            except:
+                param_name = ''
```
Shouldn't we initialize param_type also here? And specify IndexError for the exception, catching full exceptions can lead to unexpected errors.

albestro commented:
I've refactored the `parse_param` function by moving it to `Par` class and making it a `classmethod` function.

Moreover, I implemented some very basic tests for `parse_param` and I improved it.

There is also a minor fix in another test.

oaubert commented:
Awesome contribution, many thanks. I just merged it.
